integration/apidefinition_test.go

```diff
@@ -1093,3 +1095,302 @@ func TestAPIDefinition_GraphQL_ExecutionMode(t *testing.T) {
+						apiDefObj, ok := object.(*v1alpha1.ApiDefinition)
+						eval.True(ok)
+
+						return apiDefObj.Spec.GraphQL != nil &&
+							apiDefObj.Spec.GraphQL.GraphRef == newSgName &&
+							apiDefObj.Spec.GraphQL.Schema == newSchema &&
+							apiDefObj.Spec.GraphQL.Subgraph.SDL == newSDL &&
+							apiDefObj.Status.LinkedToSubgraph == newSgName
+					}),
+					wait.WithTimeout(defaultWaitTimeout),
+					wait.WithInterval(defaultWaitInterval),
+				)
+				eval.NoErr(err)
+
+				return ctx
+			}).
+		Assess("remove GraphRef from ApiDefinition",
+			func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+				eval := is.New(t)
+
+				testNs, ok := ctx.Value(ctxNSKey).(string)
+				eval.True(ok)
+
+				// Get ApiDefinition and remove GraphRef from it.
+				api := &v1alpha1.ApiDefinition{ObjectMeta: metav1.ObjectMeta{Name: testApiDef, Namespace: testNs}}
+				err := wait.For(
+					conditions.New(c.Client().Resources()).ResourceMatch(api, func(object k8s.Object) bool {
+						return true
+					}),
+					wait.WithTimeout(defaultWaitTimeout),
+					wait.WithInterval(defaultWaitInterval),
+				)
+				eval.NoErr(err)
+
+				_, err = util.CreateOrUpdate(ctx, r.Client, api, func() error {
+					api.Spec.GraphQL.GraphRef = ""
+					return nil
+				})
+				eval.NoErr(err)
+
+				// Wait for reconciliation; so that, the ApiDefinition is updated according to new SubGraph CR.
+				err = wait.For(
+					conditions.New(c.Client().Resources()).ResourceMatch(api, func(object k8s.Object) bool {
+						_, err = r.Reconcile(ctx, ctrl.Request{NamespacedName: cr.ObjectKeyFromObject(api)})
+						return err == nil
+					}),
+					wait.WithTimeout(defaultWaitTimeout),
+					wait.WithInterval(defaultWaitInterval),
+				)
+				eval.NoErr(err)
+
+				const (
+					newSDL    = "newSDL"
+					newSchema = "newSchema"
+				)
+
+				// After successful reconciliation, check that ApiDefinition CR is updated properly.
+				err = wait.For(
+					conditions.New(c.Client().Resources()).ResourceMatch(api, func(object k8s.Object) bool {
+						apiDefObj := &v1alpha1.ApiDefinition{}
+						err = r.Get(ctx, cr.ObjectKeyFromObject(api), apiDefObj)
+						if err != nil {
+							return false
+						}
+
+						return apiDefObj.Spec.GraphQL != nil &&
+							apiDefObj.Spec.GraphQL.GraphRef == "" &&
+							apiDefObj.Spec.GraphQL.Schema == newSchema &&
+							apiDefObj.Spec.GraphQL.Subgraph.SDL == newSDL &&
+							apiDefObj.Status.LinkedToSubgraph == ""
+					}),
+					wait.WithTimeout(defaultWaitTimeout),
+					wait.WithInterval(defaultWaitInterval),
+				)
+				eval.NoErr(err)
+
+				return ctx
+			}).
+		Feature()
+
+	testenv.Finish(func(ctx context.Context, c *envconf.Config) (context.Context, error) {
+		if t.Skipped() || majorTykVersion < supportedMajorTykVersion {
+			return ctx, nil
+		}
+
+		eval := is.New(t)
+		testNs, ok := ctx.Value(ctxNSKey).(string)
+		eval.True(ok)
+
+		err := r.DeleteAllOf(ctx, &v1alpha1.ApiDefinition{}, cr.InNamespace(testNs))
```

Comment:
Why are you deleting APIs and Subgraphs here? Namespace will be deleted anyway after execution.

Reply:
this is pretty useful while running tests on the local machine. Namespaces got stuck in the Terminating state easily if resources within the namespace are not deleted successfully. That's why I first delete ApiDefinitions then delete SubGraph CRs because if you first try deleting SubGraph CRs, the controller won't allow you to do that due to resource dependency.

Reply:
Makes sense